src/library/serato/seratofeature.cpp

```diff
@@ -0,0 +1,1214 @@
+    html.append(QString("<p>%1</p>").arg(summary));
+    html.append(QString("<ul>"));
+    for (const auto& item : items) {
+        html.append(QString("<li>%1</li>").arg(item));
+    }
+    html.append(QString("</ul>"));
+
+    //Colorize links in lighter blue, instead of QT default dark blue.
+    //Links are still different from regular text, but readable on dark/light backgrounds.
+    //https://bugs.launchpad.net/mixxx/+bug/1744816
+    html.append(QString("<a style=\"color:#0496FF;\" href=\"refresh\">%1</a>")
+                        .arg(refreshLink));
+    return html;
+}
+
+void SeratoFeature::refreshLibraryModels() {
+}
+
+void SeratoFeature::activate() {
+    qDebug() << "SeratoFeature::activate()";
+
+    // Let a worker thread do the parsing
+    m_databasesFuture = QtConcurrent::run(findSeratoDatabases, this);
+    m_databasesFutureWatcher.setFuture(m_databasesFuture);
+    m_title = tr("(loading) Serato");
+    //calls a slot in the sidebar model such that 'Serato (isLoading)' is displayed.
+    emit featureIsLoading(this, true);
+
+    emit enableCoverArtDisplay(true);
+    emit switchToView("SERATOHOME");
+}
+
+void SeratoFeature::activateChild(const QModelIndex& index) {
+    if (!index.isValid())
+        return;
+
+    //access underlying TreeItem object
+    TreeItem* item = static_cast<TreeItem*>(index.internalPointer());
+    if (!(item && item->getData().isValid())) {
+        return;
+    }
+
+    // TreeItem list data holds 2 values in a QList:
+    //
+    //     1. Playlist Name/Path (QString)
+    //     2. isPlaylist (boolean)
+    //
+    // If the second element is false, then the database does still have to be
+    // parsed.
+    QList<QVariant> data = item->getData().toList();
+    VERIFY_OR_DEBUG_ASSERT(data.size() == 2) {
+        return;
+    }
+    QString playlist = data[0].toString();
+    bool isPlaylist = data[1].toBool();
+
+    qDebug() << "SeratoFeature::activateChild " << item->getLabel();
+
+    if (!isPlaylist) {
+        // Let a worker thread do the parsing
+        m_tracksFuture = QtConcurrent::run(parseDatabase, static_cast<Library*>(parent())->dbConnectionPool(), item);
+        m_tracksFutureWatcher.setFuture(m_tracksFuture);
+
+        // This device is now a playlist element, future activations should
+        // treat is as such
+        data[1] = QVariant(true);
+        item->setData(QVariant(data));
+    } else {
+        qDebug() << "Activate Serato Playlist: " << playlist;
+        m_pSeratoPlaylistModel->setPlaylist(playlist);
+        emit showTrackModel(m_pSeratoPlaylistModel);
+    }
+}
+
+void SeratoFeature::onSeratoDatabasesFound() {
+    QList<TreeItem*> foundDatabases = m_databasesFuture.result();
+    TreeItem* root = m_childModel.getRootItem();
+
+    QSqlDatabase database = m_pTrackCollection->database();
+
+    if (foundDatabases.size() == 0) {
+        // No Serato databases found
+        ScopedTransaction transaction(database);
+        transaction.commit();
```

Comment:
Why do we need to start a transaction and immediately commit it without any changes? The commit starts a subsequent transaction, but it is redundant.

Reply:
oops